src/services/paras/ParasService.ts

```diff
@@ -0,0 +1,441 @@
+import { u32 } from '@polkadot/types';
+import { Option, Vec } from '@polkadot/types/codec';
+import {
+	AccountId,
+	BalanceOf,
+	BlockHash,
+	BlockNumber,
+	FundInfo,
+	ParaGenesisArgs,
+	ParaId,
+	ParaLifecycle,
+	WinningData,
+} from '@polkadot/types/interfaces';
+import { ITuple } from '@polkadot/types/types';
+import BN from 'bn.js';
+import { InternalServerError } from 'http-errors';
+
+import {
+	AuctionPhase,
+	IAuctionsCurrent,
+	ICrowdloans,
+	ICrowdloansInfo,
+	IFund,
+	ILeaseInfo,
+	ILeasesCurrent,
+	IParas,
+	LeaseFormatted,
+	ParaType,
+} from '../../types/responses';
+import { IOption, isSome } from '../../types/util';
+import { AbstractService } from '../AbstractService';
+
+// This was the orgiginal value in the rococo test net. Once the exposed metadata
+// consts makes its way into `rococo-v1` this can be taken out.
+const LEASE_PERIODS_PER_SLOT_FALLBACK = 4;
+
+export class ParasService extends AbstractService {
+	/**
+	 * Get crowdloan information for a `paraId`.
+	 *
+	 * @param hash `BlockHash` to make call at
+	 * @param paraId ID of para to get crowdloan info for
+	 */
+	async crowdloansInfo(
+		hash: BlockHash,
+		paraId: number
+	): Promise<ICrowdloansInfo> {
+		const [fund, { number }] = await Promise.all([
+			this.api.query.crowdloan.funds.at<Option<FundInfo>>(hash, paraId),
+			this.api.rpc.chain.getHeader(hash),
+		]);
+
+		if (!fund) {
+			throw new InternalServerError(
+				`Could not find funds info at para id: ${paraId}`
+			);
+		}
+
+		let fundInfo, leasePeriods;
+		if (fund.isSome) {
+			fundInfo = fund.unwrap();
+			const firstSlot = fundInfo.firstSlot.toNumber();
+			// number of lease periods this crowdloan covers
+			const leasePeriodCount = fundInfo.lastSlot.toNumber() - firstSlot + 1;
+			leasePeriods = Array(leasePeriodCount)
+				.fill(0)
+				.map((_, i) => i + firstSlot);
+		} else {
+			fundInfo = null;
+		}
+
+		const at = {
+			hash,
+			height: number.unwrap().toString(10),
+		};
+
+		return {
+			at,
+			fundInfo,
+			leasePeriods,
+		};
+	}
+
+	/**
+	 * List all available crowdloans.
+	 *
+	 * @param hash `BlockHash` to make call at
+	 * @param includeFundInfo wether or not to include `FundInfo` for every crowdloan
+	 */
+	async crowdloans(
+		hash: BlockHash,
+		includeFundInfo: boolean
+	): Promise<ICrowdloans> {
+		const [{ number }, funds] = await Promise.all([
+			this.api.rpc.chain.getHeader(hash),
+			this.api.query.crowdloan.funds.entriesAt<Option<FundInfo>, [ParaId]>(
+				hash
+			),
+		]);
+
+		let entries: IFund[];
+		if (includeFundInfo) {
```

Comment:
Yeah, now that you point it out I totally agree it is not the order the query can be made in.

From design perspective I think it could make sense to do the below query for just the storage keys, and then extract the `paraId`s by that, which I assume is a bit quicker. Then the user can get `fundInfo` with `/paras/{paraId}/crowdloan-info`. So this endpoint we be discovering who has crowdloans, but to get the actual crowdloan would be a call for each `paraId`. So, If we include the `fundInfo` here then we are slightly redundant since you can get the fundInfo from two places.
More of a nitpicky detail though and I can be convinced either - so happy with implementing your suggestion.

Reply:
We're experimenting here so I think it's fine if we have some things in two places. Maybe we can have it in both places and see what our users prefer?

Reply:
👍  I changed it here and in the docs to always include the fund info.